src/treemodel/history.js

```diff
@@ -0,0 +1,126 @@
+		 * at which the delta was added.
+		 *
+		 * @private
+		 * @member {Map} core.treeModel.History#_historyPoints
+		 */
+		this._historyPoints = new Map();
+	}
+
+	/**
+	 * Gets the number of base version which an up-to-date operation should have.
+	 *
+	 * @private
+	 * @type {Number}
+	 */
+	get _nextHistoryPoint() {
+		const lastDelta = this._deltas[ this._deltas.length - 1 ];
+
+		return lastDelta.baseVersion + lastDelta.operations.length;
+	}
+
+	/**
+	 * Adds an operation to the history.
+	 *
+	 * @param {core.treeModel.operation.Operation} operation Operation to add.
+	 */
+	addOperation( operation ) {
+		const delta = operation.delta;
+
+		// History cares about deltas not singular operations.
+		// Operations from a delta are added one by one, from first to last.
+		// Operations from one delta cannot be mixed with operations from other deltas.
+		// This all leads us to the conclusion that we could just save deltas history.
+		// What is more, we need to check only the last position in history to check if delta is already in the history.
+		if ( delta && this._deltas[ this._deltas.length - 1 ] !== delta ) {
+			const index = this._deltas.length;
+
+			this._deltas[ index ] = delta;
+			this._historyPoints.set( delta.baseVersion, index );
+		}
+	}
+
+	/**
+	 * Transforms out-dated delta by all deltas that were added to the history since the given delta's base version. In other
+	 * words, it makes the delta up-to-date with the history. The transformed delta(s) is (are) ready to be applied
+	 * to the {@link core.treeModel.Document document}.
+	 *
+	 * @param {core.treeModel.delta.Delta} delta Delta to update.
+	 * @returns {Array.<core.treeModel.delta.Delta>} Result of transformation which is an array containing one or more deltas.
+	 */
+	updateDelta( delta ) {
+		if ( delta.baseVersion === this._nextHistoryPoint ) {
+			return [ delta ];
+		}
+
+		let index = this._historyPoints.get( delta.baseVersion );
+
+		if ( index === undefined ) {
+			throw new CKEditorError( 'history-wrong-version: Cannot retrieve point in history that is a base for given delta.' );
+		}
+
+		let transformed = [ delta ];
+
+		while ( index < this._deltas.length ) {
+			const historyDelta = this._deltas[ index ];
+			let allResults = [];
+
+			for ( let deltaToTransform of transformed ) {
+				const transformedDelta = History._transform( deltaToTransform, historyDelta );
+				allResults = allResults.concat( transformedDelta );
+			}
+
+			transformed = allResults;
+			index++;
+		}
+
+		return transformed;
+	}
+
+	/**
+	 * Transforms given delta by another given delta.
+	 *
+	 * @private
+	 * @param {core.treeModel.delta.Delta} toTransform Delta to be transformed.
+	 * @param {core.treeModel.delta.Delta} transformBy Delta to transform by.
+	 */
+	static _transform( toTransform, transformBy ) {
+		return transform( toTransform, transformBy, false );
```

Comment:
For testing purposes. We have discussed it on the chat and a bit live with @reinmar.

"I want to test and I want to test whether `transform` was called and with what arguments. I see 3 solutions:
1. Test the effect of the whole loop. This sucks because it is mostly testing what `transform` does and I already have tests for this.
2. Make `transform` static member of that class, which will allow spying it. This is weird because w change the class only because of tests requirements. OTOH it would let us bring in different `transform` method but that won't happen, like, ever.
3. Move the inner `for` to other method and test whether that method got called.  This kinda solves the issue but "the other method" would probably be private and not tested."

We chose to go with solution 2.